chain-tx-enclave-next/mls/src/extras/mod.rs

```diff
@@ -329,10 +330,10 @@ mod test {
             .process_commit(commit.clone(), &proposals, &ra_verifier, 0)
             .expect_err("commit not ok");
         let ctx = member3_group.context.get_encoding();
-        // for group 3, it should look ok
+        // for group 3, it failed at GroupInfoIntegrityError
         member3_group
             .process_commit(commit.clone(), &proposals, &ra_verifier, 0)
-            .expect("commit ok");
+            .expect_err("commit not ok");
```

Comment:
member3 can detect the computed confirmation is different from the one contained in message.

Reply:
ok, then the corruption code should also update the confirmation (it could also construct the corrupted message from scratch as intended if it's easier than tweaking previously valid one)

Reply:
changed it to assert specific error branch.

Reply:
it seems not easy to construct a corrupted message with correct confirmation, unless basically copy and modify the most of the `commit_proposals` function.

Reply:
No, that's incorrect to assert and wouldn't test whether the intended functionality really works. The test case setup is:
- member1 is the affected one (and can detect it immediately from Commit, because secrets on its direct path either can't be decrypted OR don't correspond to the public values)
- member2 generates the corrupted message
- member3 can't tell Commit is wrong immediately (as the secrets shouldn't be on its path) -- and can only tell it from the nack message that discloses member1's shared secret and proof

Reply:
For the verification, member3 needs the expected public nodes which are in the updated tree -- currently in the code, one can only get the updated tree from "process commit" (ideally, there'll be some "verify_commit", so that one can revert back to the old state if later nack message shows the commit was invalid), so if the "process commit" failed for member3, it won't use the correct input (so even though this check https://github.com/crypto-com/chain/blob/199e80b371445ba644fc807179f0ad0392d9901c/chain-tx-enclave-next/mls/src/extras/mod.rs#L184 fails, as expected; it fails due to a wrong input tree... not due to a correct Nack)

Reply:
I think it's doable like this:
- re-compute transcript_hash with new commit
- get tree_hash from welcome message
- build updated GroupContext
- get `commit_secret` from committer `GroupAux`'s `pending_commit`
- gen `epoch_secrets` and `confirmation`

Reply:
I think it's doable like this:
- re-compute transcript_hash with new commit
- get tree_hash from welcome message (decrypt group info object)
- build updated GroupContext
- to compute `commit_secret`:
  - get updated public tree from welcome message
  - get `init_private_key` from commtter's `pending_commit`
  - get `tree_secret` by evolve the tree with the `init_private_key`
  - get `commit_secret` from `tree_secret`
- get `commit_secret` by apply path secrets.
- gen `epoch_secrets` and `confirmation`